fem/tbilininteg.hpp

```diff
@@ -554,6 +554,1126 @@ struct TDiffusionKernel<3,3,complex_t>
    }
 };
 
+// Mass kernel for Nedelec Finite Element
+
+// complex_t - type for the assembled data
+template <int SDim, int Dim, typename complex_t>
+struct THcurlMassKernel;
+
+// Mass kernel for Nedelec Finite Element kernel in 2D
+template <typename complex_t>
+struct THcurlMassKernel<2,2,complex_t>
+{
+   typedef complex_t complex_type;
+
+   // needed for the TElementTransformation::Result class
+   static const bool uses_Jacobians = true;
+
+   // needed for the FieldEvaluator::Data class
+   static const bool in_values  = true;
+   static const bool in_curls   = false;
+   static const bool out_values = true;
+   static const bool out_curls  = false;
+
+   // Partially assembled data type for one element with the given number of
+   // quadrature points. This type is used in partial assembly, and partially
+   // assembled action. Stores one symmetric 2 x 2 matrix per point.
+   template <int qpts>
+   struct p_asm_data { typedef TMatrix<qpts,3,complex_t> type; };
+
+   // Partially assembled data type for one element with the given number of
+   // quadrature points. This type is used in full element matrix assembly.
+   // Stores one general (non-symmetric) 2 x 2 matrix per point.
+   template <int qpts>
+   struct f_asm_data { typedef TTensor3<qpts,2,2,complex_t> type; };
+
+   template <typename IR, typename coeff_t, int NE>
+   struct CoefficientEval
+   {
+      typedef typename IntRuleCoefficient<IR,coeff_t,NE>::Type Type;
+   };
+
+   // Method used for un-assembled (matrix free) action.
+   // Jt        [M x Dim x SDim x NE] - Jacobian transposed, data member in F
+   // Q                               - CoefficientEval<>::Type
+   // q                               - CoefficientEval<>::Type::result_t
+   // val_qpts [M x SDim x NC x NE]  - in/out data member in R
+   //
+   // val_qpts = (w/det(J)) adj(J) adj(J)^t val_qpts
+   template <typename T_result_t, typename Q_t, typename q_t,
+             typename S_data_t>
+   static inline MFEM_ALWAYS_INLINE
+   void Action(const int k, const T_result_t &F,
```

Comment:
I don't understand why the mass Kernel would have this form: val_qpts = (w/det(J)) adj(J) adj(J)^t val_qpts. Isn't there a confusion with the Laplacian or some other operator involving derivatives?